sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -145,39 +159,59 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       distinctNumPreShufflePartitions.length == 1,
       "There should be only one distinct value of the number pre-shuffle partitions " +
         "among registered Exchange operator.")
-    val numPreShufflePartitions = distinctNumPreShufflePartitions.head
 
     val partitionStartIndices = ArrayBuffer[Int]()
-    // The first element of partitionStartIndices is always 0.
-    partitionStartIndices += 0
+    val partitionEndIndices = ArrayBuffer[Int]()
 
-    var postShuffleInputSize = 0L
+    def nextStartIndex(i: Int): Int = {
+      if (i == validPartitions.length - 1) i else i + 1
+    }
 
-    var i = 0
-    while (i < numPreShufflePartitions) {
-      // We calculate the total size of ith pre-shuffle partitions from all pre-shuffle stages.
-      // Then, we add the total size to postShuffleInputSize.
-      var nextShuffleInputSize = 0L
+    def partitionSize(partitionId: Int): Long = {
+      var size = 0L
       var j = 0
       while (j < mapOutputStatistics.length) {
-        nextShuffleInputSize += mapOutputStatistics(j).bytesByPartitionId(i)
+        val statistics = mapOutputStatistics(j)
+        size += statistics.bytesByPartitionId(partitionId)
         j += 1
       }
+      size
+    }
+    var j = 0
+    val firstStartIndex = validPartitions(j)
+
+    partitionStartIndices += firstStartIndex
+
+    var postShuffleInputSize = partitionSize(firstStartIndex)
+
+    var i = firstStartIndex
+    j = nextStartIndex(j)
+    var nextIndex = validPartitions(j)
+    var k = 1
+    while (k < validPartitions.length) {
```

Comment:
There are so many temp variables that make the code hard to follow here but `k` is not used. I think `validPartitions.foreach { _ => ... }` could make it easier.